### ISSUE TITLE
Session initialization takes long time on NVIDIA RTX type GPUs

### ISSUE THREAD

**Issue by itsjj00** [microsoft/onnxruntime#9990]
Taking over 20 min on CreateSession with GPU option on NVIDIA RTX A4000 Laptop GPU and RTX A6000

**Urgency**
This issue causes our ML functions not working properly in our customer environment.

**System information**
- OS Platform : Windows 10  C++
- ONNX Runtime installed from (source or binary):  
- ONNX Runtime version: 1.2 native
- Visual Studio version (if applicable): VS 2017
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:  CUDA 10.1,  cuDNN 7.6.5
- GPU model and memory:

**To Reproduce**
This is a "common" code set:
	m_useGPU = useGPU;
	g_ort = OrtGetApiBase()->GetApi(ORT_API_VERSION);
	CheckStatus(g_ort->CreateEnv(ORT_LOGGING_LEVEL_WARNING, "TEST", &m_env));
	// initialize session options 
	CheckStatus(g_ort->CreateSessionOptions(&m_session_options));
	g_ort->SetIntraOpNumThreads(m_session_options, numProcessors);
	g_ort->SetSessionGraphOptimizationLevel(m_session_options, ORT_ENABLE_EXTENDED);
	if (m_useGPU)
		OrtSessionOptionsAppendExecutionProvider_CUDA(m_session_options, 0);
	m_bSessionInitialized = CheckStatus(g_ort->CreateSession(m_env, m_model_path, m_session_options, &m_session));

- In this code with RTX GPUs, with m_useGPU = TRUE, the CreateSession() line takes about 30 mins, before the GPU was used to process the model.  There is no problem in house with version of Nvidia Quadro  GPU devices, which takes < 1 min.
- It does not appear that the ONNX model is the problem. The problem can be duplicated easily with Onyx_CXX_Api_Sample() code which use sample model  squeezenet.onnx

** Loaded related libraries **
cudart64_101.dll
cufft64_10.dll
onnxruntime.dll
cublas64_10.dll
cudnn64_7.dll
curand64_10.dll
cublasLt64_10.dll

We urgently needs advice to resolve this customer issue which apparently related to supporting of newer version of Nvidia drivers in the ONNXRUNTIME library.   Please let me know any additional information you may need to help us resolve this issue.
THANKS,
Jing

**Comment by jywu-msft** [comment]
Is onnxruntime built from source or using a release binary package?

**Comment by askhade** [comment]
*ONNX Runtime version: 1.2 native* is very old. Are you sure it is 1.2?

Have you tried with the latest ort package? One possible reason is the ort binary you are using was not built with the compute capability applicable to your GPU because of which jit compilation actually takes place during CUDA initialization and hence the long latency. Having said that 20 mins is still very high. I would except the session initialization latency to be ~60 secs due to jit compilation as opposed to 1-2 secs. 

You can run 
```
 cuobjdump -sass libonnxruntime_providers_cuda.so  | grep "arch ="
```
and check whether ort binary includes the right sm_arch applicable to your gpu. If it does then we can rule out jit compilation.

**Comment by jywu-msft** [comment]
sm_86 is what's required for rtx4000,6000's, and I don't believe onnxruntime is built with that architecture enabled by default.
if one builds from source, you can add support for a specific cuda arch via CMAKE_CUDA_ARCHITECTURES
i.e. add this build option to the build.bat command 
--cmake_extra_defines CMAKE_CUDA_ARCHITECTURES=86

**Comment by hariharans29** [comment]
@jywu-msft: But we build with sm_80 right ? Shouldn't this binary code be useable in CC 8.6 cards because of Nvidia's one minor revision binary compatibility guarantee (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#binary-compatibility) ? Or is my understanding of this incorrect ?

**Comment by itsjj00** [comment]
Thanks all for the reply.  We do not build onnxruntime.dll, just take from the package and use it in our product on the trained CNN Model.   Our released product runs on Cuda 10.1, onnxruntime v1.10, cudnn v7 combination which apparently does not work well with the newer NVIDIA RTX cards.  I am testing a higher set of library combination according to the chart in tis link https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#gpu_mem_limit.  Hope the higher combination will solve the RTX issue.  Also I will need to understand what is the appropriate sm and compute numbers to set that will work for a wide ranges of GPU cards, ours is set to sm_50, compute_50 is it OK?  I  am also worry of the higher combination will drop the support of lower/older version of GPU cards and cause complaint from the other end.

**Comment by jywu-msft** [comment]
Can you confirm whether you're using onnxruntime 1.1 or 1.10 (1.10 was released just a few days ago)
I ask because you refer to CUDA 10.1, which is a very old version. 
The 1.10 release packages depend on CUDA 11.x

**Comment by itsjj00** [comment]
> Can you confirm whether you're using onnxruntime 1.1.0 or 1.10.0 (1.10.0 was released just a few days ago) I ask because you refer to CUDA 10.1, which is a very old version. The 1.10.0 release packages depend on CUDA 11.x
You are right, the onnxruntime version in our release is the old 1.1.0 which works with CUDA 10.1.

**Comment by itsjj00** [comment]
Just finished a test environment with onnxruntime v 1.17,  Cuda v 11.0, Cudnn v8.02.39.  I am pleased that my unit test gives impressive time of 8 sec.  It looks the Cudnn version 8 is the one needed to support RTX GPU.   Besides cudnn64_8.dll, I also included cudnn_cnn_infer64_8.dll and cudnn_ops_infer64_8.dll  
In CUDA property, I set sm_50, compute_50.  Should I set a higher sm and compute value?  
THANKS.

**Comment by jywu-msft** [comment]
you should set the cuda arch to what your gpu device(s) can support.

**Comment by snnn** [comment]
This issue has been automatically closed as 'not planned' because it has been marked as 'stale' for more than 30 days without activity. If you believe this is still an issue, please feel free to reopen it.
